### ISSUE TITLE
Specify supported Rust versions?

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/cargo#837]
**TL;DR:** building Servo with the wrong Rust version is a terrible experience and easy to do by mistake, please help us make it harder.

---

Servo uses its own Rust snapshot, with each Servo commit targeting a specific commit. `mach build` in Servo downloads a snapshot if necessary, then calls `cargo build` with the `PATH` and `LD_LIBRARY_PATH` environment variables set so that Cargo uses the Rust snapshot.

However, people (myself included) regularly just type `cargo build` out of habit. Cargo will then use the Rust version installed system-wide if there is one, which might be a nigthly or a release, but is probably incompatible with Servo’s snapshot. Cargo will then start re-building everything (since we’ve changed Rust versions) before (probably) hitting a build error several minutes later. Even if the mistake it caught early and the developer is not utterly confused about what’s going on, the target directory has already been wiped and the developer is punished by having the next `mach build` do a full build from scratch, which takes a while.

We’ve filed https://github.com/servo/servo/issues/3957 to do something about this, but I believe this requires some help from Cargo. Would it make sense for libraries to specify in `Cargo.toml` a range of Rust versions they support?

@Manishearth also suggests
> Cargo.lock should probably store the rustc version it was compiled with (autogenerated, yay), Cargo.toml can store how fuzzy the rustc version is allowed to be.

**Comment by alexcrichton** [comment]
I think that for now there's not a lot actionable in this ticket, I agree that we'll definitely want to re-evaluate post-1.0, but for now I don't think cargo is going to enter the business of supporting various Rust versions as it's just too unstable to track currently.

**Comment by whitequark** [comment]
@alexcrichton now that we're post 1.0, can we reopen this?

**Comment by alexcrichton** [comment]
Nowadays discussion for this seems best suited at https://github.com/rust-lang/rfcs/pull/1953 perhaps?

**Comment by shaleh** [comment]
So you have to use nightly to really use Rust. And nightly breaks things. It would be really nice to specify "known working with 10.22.2017". otherwise a user who pulls a project never knows if it will build.

**Comment by SimonSapin** [comment]
@shaleh You can now have [a `rust-toolchain` file](https://github.com/rust-lang-nursery/rustup.rs/#the-toolchain-file) that contains `nightly-2017-10-22`. It’s a bit stronger than just “known to work”, though: it’ll make rustup select that exact version and download it if necessary, unless [overridden](https://github.com/rust-lang-nursery/rustup.rs/#override-precedence).

**Comment by shaleh** [comment]
@SimonSapin thank you. That works.
